### ISSUE TITLE
Raspberry Pi Support

### ISSUE THREAD

**Issue by coderpro1234-2** [Mrmayman/quantum-launcher#1]
Could you add a build that supports the raspberry pi's architecture (aarch64), or if not, add a guide to build it yourself for your specific architecture.

**Comment by Mrmayman** [comment]
I made an experimental build but I can't test it since I don't have a raspberry pi. Do you have a raspberry pi? Can you test the builds I send you?

If you can test the build then please join our discord https://discord.com/invite/bWqRaSXar5 and ping me @Mrmayman in discord

**Comment by Ash177048** [comment]
I have Raspberry Pi 4 and Pi 5 you can send me the build I can test them for you.

**Comment by Mrmayman** [comment]
https://drive.google.com/file/d/1fYh7QbWhXqMX389263TYll57yfwMlONz/view?usp=sharing

Here is a build, can you test it and see how far it goes? Just extract the zip and open the executable inside (preferably with terminal)
Note: I put it on google drive because im having issues uploading here

**Comment by coderpro1234-2** [comment]
After trying it (from the google drive link), it runs, but the textures of the menus are all messed up (SDL / OpenGL issue?), and it fails at trying to download java, with it stuck at saying 'Starting...' I might be able to get a screen recording later, if it helps diagnose some of the issues.

**Comment by Mrmayman** [comment]
hmm interesting. The java downloading thing is expected as I haven't gotten that to fully work on arm yet.

But the texture glitch is surprising. Never expected that, can you send a screenshot (or a video)?

**Comment by coderpro1234-2** [comment]
Here is a screen recording showing the graphical glitches:

https://github.com/user-attachments/assets/66a7f4a1-5a16-40ce-b11e-721e11a8be7d

**Comment by Mrmayman** [comment]
[quantum_launcher_aarch64_linux_BETA_BUILD.zip](https://github.com/user-attachments/files/18301074/quantum_launcher_aarch64_linux_BETA_BUILD.zip)

@coderpro1234-2 I think I fixed the Java Installing problem. Can you please test this? (I don't know how to fix the texture glitch though)

**Comment by coderpro1234-2** [comment]
With the new build, the textures are still glitched, but it now downloads java correctly. It would be nice to have a progress bar (or maybe a text showing percentage done due to graphical issues) to show how much is downloaded, as it took a while, and I wasn't sure if it was working or not. After it downloaded java, it started up Minecraft, but then crashes on a wrong architecture for liblwjgl.so. I will upload the log that it spits out after this.

**Comment by coderpro1234-2** [comment]
Launching Minecraft (4 January 2025)
OS: linux

! Jan 04, 2025 1:38:37 PM net.minecraft.launchwrapper.LogWrapper log
! INFO: Using tweak class name net.minecraft.launchwrapper.VanillaTweaker
! Jan 04, 2025 1:38:37 PM net.minecraft.launchwrapper.LogWrapper log
! INFO: Launching wrapped minecraft
27 achievements
174 recipes
Turning of ImageIO disk-caching
Loading current icons for window from: /home/oscar/.config/QuantumLauncher/assets/dir/icons/icon_16x16.png and /home/oscar/.config/QuantumLauncher/assets/dir/icons/icon_32x32.png
! javax.imageio.IIOException: Can't read input file!
! 	at javax.imageio.ImageIO.read(ImageIO.java:1302)
! 	at net.minecraft.launchwrapper.injector.VanillaTweakInjector.loadIcon(VanillaTweakInjector.java:130)
! 	at net.minecraft.launchwrapper.injector.VanillaTweakInjector.loadIconsOnFrames(VanillaTweakInjector.java:107)
! 	at net.minecraft.launchwrapper.injector.VanillaTweakInjector.inject(VanillaTweakInjector.java:94)
! 	at net.minecraft.client.Minecraft.main(SourceFile:2115)
! 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
! 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Setting gameDir to: /home/oscar/.config/QuantumLauncher/instances/1.0/.minecraft
! 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
! 	at java.lang.reflect.Method.invoke(Method.java:498)
! 	at net.minecraft.launchwrapper.Launch.launch(Launch.java:57)
! 	at net.minecraft.launchwrapper.Launch.main(Launch.java:18)
! Exception in thread "Minecraft main thread" java.lang.UnsatisfiedLinkError: /home/oscar/.config/QuantumLauncher/instances/1.0/libraries/natives/liblwjgl.so: /home/oscar/.config/QuantumLauncher/instances/1.0/libraries/natives/liblwjgl.so: wrong ELF class: ELFCLASS32 (Possible cause: architecture word width mismatch)
! 	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
! 	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1934)
! 	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1850)
! 	at java.lang.Runtime.loadLibrary0(Runtime.java:843)
! 	at java.lang.System.loadLibrary(System.java:1136)
! 	at org.lwjgl.Sys$1.run(Sys.java:73)
! 	at java.security.AccessController.doPrivileged(Native Method)
! 	at org.lwjgl.Sys.doLoadLibrary(Sys.java:66)
! 	at org.lwjgl.Sys.loadLibrary(Sys.java:95)
! 	at org.lwjgl.Sys.<clinit>(Sys.java:112)
! 	at org.lwjgl.opengl.Display.<clinit>(Display.java:135)
! 	at net.minecraft.client.Minecraft.a(SourceFile:187)
! 	at net.minecraft.client.Minecraft.run(SourceFile:644)
! 	at java.lang.Thread.run(Thread.java:750)

**Comment by coderpro1234-2** [comment]
Unless there is a version for aarch64, or there is a way to compile it (if the source code is avaliable), the only way may be to emulate it via qemu?

**Comment by Mrmayman** [comment]
there is a version for aarch64 i'm pretty sure. Thanks for the information, i'll try to implement downloading the correct file for the architecture

**Comment by Mrmayman** [comment]
[quantum_launcher_armbeta.zip](https://github.com/user-attachments/files/18306578/quantum_launcher_armbeta.zip)

@coderpro1234-2 Can you try this one? (please delete and remake your instance)
Btw there is a progress text and bar, it's somehow broken due to the rendering issues (I still dont know how to fix it though)

**Comment by coderpro1234-2** [comment]
With the new version, it still has the wrong architecture for liblwjgl.so (being ELFCLASS32, or x86 architecture). If you could fix this, and make sure that the other .so files are also the correct architecture for the Raspberry Pi, then launching Minecraft. Once Minecraft is able to launch correctly, then I can help give ideas as to why the rendering is not working. If you need I can provide the logs for this crash (though they are almost identical to the previous ones), and I can provide a screen recording showing the issue.

**Comment by coderpro1234-2** [comment]
After looking through the debug info when launching through the terminal (which is what I should have been doing all alone), it seems that your code is downloading an arm64 version of jinput, but no of lwjgl itself. Could you update the code  as to ALSO (so don't change the jinput) download lwjgl for arm64. I will post a log of what the terminal outputted when redownloading the Minecraft version and java.

**Comment by coderpro1234-2** [comment]
[info] You can run ./quantum_launcher --help to see the possible command line arguments
[info] Welcome to QuantumLauncher! This terminal window just outputs some debug info. You can ignore it.
[info] Starting up the launcher...
[error] Could not check for updates: update check error: current version is ahead of latest version! dev build?
[info] Started creating instance.
[info] Initializing instance folder.
[info] Started downloading version manifest JSON.
[info] Started downloading version details JSON.
[info] Downloading game jar file.
[info] Starting download of libraries.
[info] Patching net.java.jinput:jinput-platform:2.0.5
[info] Patching net.java.jinput:jinput:2.0.5
[info] Patching net.java.jutils:jutils:1.0.0
[info] Downloading https://libraries.minecraft.net/net/minecraft/launchwrapper/1.5/launchwrapper-1.5.jar
[info] Downloading https://libraries.minecraft.net/net/sf/jopt-simple/jopt-simple/4.5/jopt-simple-4.5.jar
[info] Downloading https://libraries.minecraft.net/org/lwjgl/lwjgl/lwjgl_util/2.9.0/lwjgl_util-2.9.0.jar
[info] Downloading https://libraries.minecraft.net/net/java/jutils/jutils/1.0.0/jutils-1.0.0.jar
[info] Downloading https://libraries.minecraft.net/org/ow2/asm/asm-all/4.1/asm-all-4.1.jar
[info] Downloading https://libraries.minecraft.net/org/lwjgl/lwjgl/lwjgl/2.9.0/lwjgl-2.9.0.jar <- Bad, needs to be for arm64 / aarch64 (both the same thing).
[info] Downloading https://github.com/theofficialgman/lwjgl3-binaries-arm64/raw/lwjgl-2.9.4/jinput-2.0.5.jar <- Good
[info] Finished creating instance: 1.0
[info] Installing Java: java_8
[info] Started installing java_8
[info] Installing Java
[info] Finished installing java_8
[info] Finished installing Java java_8
[info] Java args: ["-Xss1M", "-Dminecraft.launcher.brand=minecraft-launcher", "-Dminecraft.launcher.version=2.1.1349", "-Djava.library.path=/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/natives", "-Xmx2048M", "-cp", "/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/net/minecraft/launchwrapper/1.5/launchwrapper-1.5.jar:/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/net/sf/jopt-simple/jopt-simple/4.5/jopt-simple-4.5.jar:/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/org/ow2/asm/asm-all/4.1/asm-all-4.1.jar:/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/net/java/jinput/jinput/2.0.5/jinput-2.0.5.jar:/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/net/java/jutils/jutils/1.0.0/jutils-1.0.0.jar:/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/org/lwjgl/lwjgl/lwjgl/2.9.0/lwjgl-2.9.0.jar:/home/oscar/.config/QuantumLauncher/instances/1.0/libraries/org/lwjgl/lwjgl/lwjgl_util/2.9.0/lwjgl_util-2.9.0.jar:/home/oscar/.config/QuantumLauncher/instances/1.0/.minecraft/versions/1.0/1.0.jar", "net.minecraft.launchwrapper.Launch"]

[info] Game args: ["test", "${auth_session}", "--gameDir", "/home/oscar/.config/QuantumLauncher/instances/1.0/.minecraft", "--assetsDir", "/home/oscar/.config/QuantumLauncher/assets/null"]

[info] Game exited with status: exit status: 0

**Comment by coderpro1234-2** [comment]
*HUGE UPDATE* I don't know what the .jar library files do, and unless they are compiled to the .so files, the .so files are the ones that need to be replaced with aarch64 versions. If you install the 'liblwjgl-java' package using apt, and copy the .so files that it install from /usr/lib/jni to the Minecraft folder where the .so are located (and rename the libjinput.so to libjinput-linux.so), it runs perfectly fine. You may want to find where there are located on the raspbian / debian / java servers, and use those instead on the ones that come from Mojang's servers (I don't know if they do, I am just assuming).

**Comment by coderpro1234-2** [comment]
*EVEN BIGGER UPDATE* It seems that 1.0 is broken, as when running 1.10, it downloads the correct file, and runs perfectly fine without any modification to the files by me. Anything >= 1.15 (as that is all that I have tried so far) seems to be unable to find some of the libraries, and fails. I can give some of the files that it missed, and you may want to create a little test program that goes through every Minecraft version, and see if it can find the correct library files, so you can find alternate sources for the ones that no longer exist on the servers.

**Comment by Mrmayman** [comment]
wow, thanks for the detailed information! I'll try to fix the library downloading :)

**Comment by coderpro1234-2** [comment]
I will try to tomorrow, though if you create a test app to go through every Minecraft version, and see if you can contact the servers for the libraries (not download them, at that would take forever), and report the ones that are missing, that would work better, as some version are missing some, while other versions are missing others, and I may miss a version that had a broken library link.

**Comment by Mrmayman** [comment]
[quantum_launcher_arm64_beta.zip](https://github.com/user-attachments/files/18309512/quantum_launcher_arm64_beta.zip)

@coderpro1234-2 can you please try this? I made many improvements to the library downloading code. Btw no need to list missing libraries if it's too hard. I'll deal with that

**Comment by coderpro1234-2** [comment]
(Note: I may miss some missing files, due to unable to check every version as that would take forever.) With the new version, 1.0 downloads all of the correct files and works without any file modifications. 1.10 works (no detail as it worked in the previous version). 1.15 is unable to find a file (https://github.com/theofficialgman/lwjgl3-binaries-arm64/raw/lwjgl-3.2.1/lwjgl-core-natives-linux-arm64.jar, maybe use lwjgl-core.jar or lwjgl-natives-linux-arm64.jar). 1.20 can find all of the necessary libraries, but gives an error saying 'GLFW error 65443: EGL: Fail to create context: Arguments are inconsistent. Please make sure you have up-to-date drivers (see aka.ms/mcdriver for instructions).' 1.21.4 (latest version) downloads all of the necessary libraries, but gives the same error as 1.20. Alpha 1.1.1 (as a test for a older version) downloads fine (which means that omniarchive works), and plays fine as well.

**Comment by coderpro1234-2** [comment]
Error Log:
Launching Minecraft (6 January 2025)
OS: linux

  [09:18:25:Datafixer Bootstrap.com.mojang.datafixers.DataFixerBuilder] [INFO] 243 Datafixer optimizations took 677 milliseconds
  [09:18:31:Render thread.alc] [INFO] [STDERR]: [LWJGL] [ERROR] Incompatible Java and native library versions detected.
Possible reasons:
	a) -Djava.library.path is set to a folder containing shared libraries of an older LWJGL version.
	b) The classpath contains jar files of an older LWJGL version.
Possible solutions:
	a) Make sure to not set -Djava.library.path (it is not needed for developing with LWJGL 3) or make
	   sure the folder it points to contains the shared libraries of the correct LWJGL version.
	b) Check the classpath and make sure to only have jar files of the same LWJGL version in it.
  [09:18:31:Render thread.com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService] [INFO] Environment: Environment[sessionHost=https://sessionserver.mojang.com, servicesHost=https://api.minecraftservices.com, name=PROD]
  [09:18:31:Render thread.flk] [INFO] Setting user: test
  [09:18:31:Render thread.alc] [INFO] [STDERR]: [LWJGL] [ERROR] Incompatible Java and native library versions detected.
Possible reasons:
	a) -Djava.library.path is set to a folder containing shared libraries of an older LWJGL version.
	b) The classpath contains jar files of an older LWJGL version.
Possible solutions:
	a) Make sure to not set -Djava.library.path (it is not needed for developing with LWJGL 3) or make
	   sure the folder it points to contains the shared libraries of the correct LWJGL version.
	b) Check the classpath and make sure to only have jar files of the same LWJGL version in it.
  [09:18:32:Render thread.flk] [INFO] Backend library: LWJGL version 3.3.3+5
  [09:18:33:Render thread.alc] [INFO] [STDERR]: [LWJGL] [ERROR] Incompatible Java and native library versions detected.
Possible reasons:
	a) -Djava.library.path is set to a folder containing shared libraries of an older LWJGL version.
	b) The classpath contains jar files of an older LWJGL version.
Possible solutions:
	a) Make sure to not set -Djava.library.path (it is not needed for developing with LWJGL 3) or make
	   sure the folder it points to contains the shared libraries of the correct LWJGL version.
	b) Check the classpath and make sure to only have jar files of the same LWJGL version in it.

**Comment by coderpro1234-2** [comment]
After testing a few more of the older versions (versions below Minecraft 1.0), they all work perfectly fine. I did find that one of the items from omniarchive shows the url, instead of the name, and the correct name is displayed above it (see picture provided). 
![QuantumLauncherBug](https://github.com/user-attachments/assets/cc234b86-85ec-4f40-89cf-950fb1ff99c5)

**Comment by Mrmayman** [comment]
Ok, i will try to fix the 1.15 and the glfw error.

Btw that url thing in the version list is a bug, thanks for reporting it

**Comment by coderpro1234-2** [comment]
WIth the GLFW error, it might involve forcing Minecraft (and java for that matter) to think that the opengl version is higher than it actually is. This can be achieved by adding MESA_GL_VERSION_OVERRIDE=x to the front of the command to launch the Minecraft jar file, where x is the opengl version (4.3 might be a good value). This might fix the issue, but I don't know how Minecraft works, and if it will fix when this is applied.

**Comment by coderpro1234-2** [comment]
when adding MESA_GL_VERSION_OVERRIDE=x to the start of the command to run quantum launcher, instead of giving the error, the game crashes, with a 'java.lang.UnsatisfiedLinkError'. I will post a copy of the crash report (the more detailed one that you get by disabling logging) in the next comment.

**Comment by coderpro1234-2** [comment]
---- Minecraft Crash Report ----
// Ouch. That hurt :(

Time: 2025-01-07 08:33:26
Description: Initializing game

java.lang.UnsatisfiedLinkError: 'long org.lwjgl.system.ThreadLocalUtil.nsetupEnvData(int)'
	at org.lwjgl.system.ThreadLocalUtil.nsetupEnvData(Native Method)
	at org.lwjgl.system.ThreadLocalUtil.setupEnvData(ThreadLocalUtil.java:164)
	at org.lwjgl.system.ThreadLocalUtil.setCapabilities(ThreadLocalUtil.java:178)
	at org.lwjgl.opengl.GL.setCapabilities(GL.java:216)
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:475)
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:311)
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:295)
	at fey.<init>(SourceFile:119)
	at gmz.a(SourceFile:21)
	at flk.<init>(SourceFile:504)
	at net.minecraft.client.main.Main.main(SourceFile:250)


A detailed walkthrough of the error, its code path and all known details is as follows:

-- Head --
Thread: Render thread
Stacktrace:
	at org.lwjgl.system.ThreadLocalUtil.nsetupEnvData(Native Method)
	at org.lwjgl.system.ThreadLocalUtil.setupEnvData(ThreadLocalUtil.java:164)
	at org.lwjgl.system.ThreadLocalUtil.setCapabilities(ThreadLocalUtil.java:178)
	at org.lwjgl.opengl.GL.setCapabilities(GL.java:216)
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:475)
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:311)
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:295)
	at fey.<init>(SourceFile:119)
	at gmz.a(SourceFile:21)
	at flk.<init>(SourceFile:504)

-- Initialization --
Details:
	Modules: 
Stacktrace:
	at net.minecraft.client.main.Main.main(SourceFile:250)

-- System Details --
Details:
	Minecraft Version: 1.21.4
	Minecraft Version ID: 1.21.4
	Operating System: Linux (aarch64) version 6.6.62+rpt-rpi-v8
	Java Version: 21.0.5, Amazon.com Inc.
	Java VM Version: OpenJDK 64-Bit Server VM (mixed mode, sharing), Amazon.com Inc.
	Memory: 195035136 bytes (186 MiB) / 353370112 bytes (337 MiB) up to 2147483648 bytes (2048 MiB)
	CPUs: 4
	Processor Vendor: ARM
	Processor Name: Raspberry Pi 5 Model B Rev 1.0
	Identifier: ARM Family 8 Model Raspberry Pi 5 Model B Rev 1.0 Stepping r4p1
	Microarchitecture: unknown
	Frequency (GHz): -0.00
	Number of physical packages: 1
	Number of physical CPUs: 4
	Number of logical CPUs: 4
	Graphics card #0 name: unknown
	Graphics card #0 vendor: unknown
	Graphics card #0 VRAM (MiB): 0.00
	Graphics card #0 deviceId: unknown
	Graphics card #0 versionInfo: unknown
	Virtual memory max (MiB): 4482.51
	Virtual memory used (MiB): 1191.01
	Swap memory total (MiB): 512.00
	Swap memory used (MiB): 0.00
	Space in storage for jna.tmpdir (MiB): <path not set>
	Space in storage for org.lwjgl.system.SharedLibraryExtractPath (MiB): <path not set>
	Space in storage for io.netty.native.workdir (MiB): <path not set>
	Space in storage for java.io.tmpdir (MiB): available: 208036.31, total: 239563.98
	Space in storage for workdir (MiB): available: 208036.31, total: 239563.98
	JVM Flags: 2 total; -Xss1M -Xmx2048M
	Launched Version: 1.21.4
	Launcher name: minecraft-launcher
	Backend library: LWJGL version 3.3.3+5
	Backend API: Unknown
	Window size: <not initialized>
	GFLW Platform: wayland
	GL Caps: Using framebuffer using OpenGL 3.2
	GL debug messages: <disabled>
	Is Modded: Probably not. Client jar signature and brand is untouched
	Universe: 404
	Type: Client (map_client.txt)
	Locale: en_AU
	System encoding: UTF-8
	File encoding: UTF-8
	CPU: <unknown>

**Comment by coderpro1234-2** [comment]
After running Quantum Launcher and Minecraft under X11 (comared to the default of Wayland Labwc), the icon of Quantum Launcher works (Labwc doesn't and shows a no icon at all), and instead of a 'GLFW error 65443: EGL: Fail to create context: Arguments are inconsistent' error, it gives 'GLFW error 65543: GLX: Failed to create context: GLXBadFBConfig'

**Comment by coderpro1234-2** [comment]
It seems that the crash (java.lang.UnsatisfiedLinkError) is due to the downloaded library not having that function, maybe due to it being to old, or some other reason. You may want to change you code to download a different file when trying to download this one.

**Comment by Mrmayman** [comment]
[quantum_launcher_aarch64_beta.zip](https://github.com/user-attachments/files/18356990/quantum_launcher_aarch64_beta.zip)

Can you test this version with and without the mesa override? @coderpro1234-2 

There was one native library for 1.20/1.21 which was the wrong version (it's not glfw though). I've fixed that

**Comment by coderpro1234-2** [comment]
When running the latest version WITHOUT the mesa overrides, it still gives the same issue ('GLFW error 65443: EGL') on wayland labwc (I will try on x11 later). If I run it WITH the mesa overrides, it gives the same crash as the previous version.

**Comment by coderpro1234-2** [comment]
After doing some digging, It seems that the windows version of the lwjgl jar (and the x86_64 linux version) have a nsetupEnvData function, while the aarch64 linux version doesn't. This is an issue as when minecraft calls the different function setupEnvData, that function calls nsetupEnvData, which doesn't exist, and crashes. If there is a way to compile the lwjgl jar file for aarch64 (into a .so file) that contains that function, it should get further into booting minecraft.

**Comment by coderpro1234-2** [comment]
When running release v0.3.1, it doesn't launch at all, and gives the error WaylandError(Connection(NoWaylandLib)). This link might help with fixing it: [https://github.com/iced-rs/iced/issues/2385](https://github.com/iced-rs/iced/issues/2385).

**Comment by Mrmayman** [comment]
Hmm this looks like a build system issue (wayland native libraries)? I'll ask apicalshark to look into this (they take care of the build system)

As for the `nsetupEnvData` issue, i'm a bit busy at the moment, i'll take a look at that in the future when i get time (maybe compile a custom native lib)

**Comment by Mrmayman** [comment]
@coderpro1234-2 please try this version and tell me if it works

[quantum_launcher_armbeta.zip](https://github.com/user-attachments/files/18723715/quantum_launcher_armbeta.zip)

**Comment by coderpro1234-2** [comment]
When running the beta version you gave me, it now launches, but has the same GLFW errors as before. If you force opengl to be 4.3, IT WORKS!!!! It works fine, without the nsetupEnvData error (maybe due to the file being updated?). Other than the graphical issues on the launcher (try searching 'iced rust raspberry pi' into google, it will give you some results that might help), it works perfectly (and at 60fps as well!).

**Comment by Mrmayman** [comment]
Oh wow! Nice to see it works

Did you use the `MESA_GL_VERSION_OVERRIDE` from earlier to force opengl to be 4.3? I might add an auto setup for that if necessary.
I added some previously missing arguments for searching of native libraries.

I'll also look into fixing the graphical issues as well when i get, so stay tuned for beta builds later!
Thanks for the work so far!

**Comment by coderpro1234-2** [comment]
Yes, I did use the `MESA_GL_VERSION_OVERRIDE` from before (full execution is `MESA_GL_VERSION_OVERRIDE=4.3 ./quantum-launcher`.

Also thank YOU for the work so far, as all I have done is report the issues, and you have done that hard work of finding the bug in the code, and fixing it (though, if you enjoy coding as much as I do, it is still very fun to do so).

**Comment by Mrmayman** [comment]
A few more things:

can you find out which is the minimum version that needs the `MESA_GL_VERSION_OVERRIDE`? Is it 1.20 or an earlier version?

And can you try `WGPU_BACKEND=gl ./quantum_launcher` and `WGPU_BACKEND=vulkan ./quantum_launcher` and see if either of them fixes the graphical glitches?

**Comment by coderpro1234-2** [comment]
The minimum version for `MESA_GL_VERSION_OVERRIDE` is 3.3 (at least on labwc), anything below gives the GLFW error.

When running with `WGPU_BACKEND=vulkan`, it displays as before (most likely actual backend). If I run under `WGPU_BACKEND=gl`, it doesn't have any of the graphical glitches, but the highlights on the boxes are missing (can provide screenshots / screen recording if it helps diagnose the issue).

**Comment by coderpro1234-2** [comment]
Do you need any more information?

**Comment by coderpro1234-2** [comment]
The new version look amazing. It also displays properly on raspberry pi, both using `vulkan` and `gl` as the renderer. Mods also work fine. Anything below 1.12.2 (including 1.12.2) works fine, without needing any `MESA_GL_VERSION_OVERRIDE` flags. 1.13 - 1.15 crash with an error saying missing library `liblwjgl_stb.so` 1.16 gives an `SIGSEGV` error in the Java Runtine Environment (will post the full crash log for 1.15 and 1.16 after this post). 1.17 and beyond require `MESA_GL_VERSION_OVERRIDE` flag to be set.

**Comment by coderpro1234-2** [comment]
1.15:
```
---- Minecraft Crash Report ----
// This doesn't make any sense!

Time: 4/04/25 10:59 AM
Description: Initializing game

java.lang.UnsatisfiedLinkError: Failed to locate library: liblwjgl_stb.so
	at org.lwjgl.system.Library.loadSystem(Library.java:146)
	at org.lwjgl.stb.LibSTB.<clinit>(LibSTB.java:17)
	at org.lwjgl.stb.STBImage.<clinit>(STBImage.java:128)
	at cxo.a(SourceFile:122)
	at cxo.a(SourceFile:98)
	at cxo.a(SourceFile:89)
	at eaz.<init>(SourceFile:29)
	at dbl.a(SourceFile:2409)
	at yt.a(SourceFile:45)
	at eal.a(SourceFile:67)
	at yq.a(SourceFile:33)
	at dbp.a(SourceFile:665)
	at dbl.<init>(SourceFile:439)
	at net.minecraft.client.main.Main.main(SourceFile:164)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Render thread
Stacktrace:
	at org.lwjgl.system.Library.loadSystem(Library.java:146)
	at org.lwjgl.stb.LibSTB.<clinit>(LibSTB.java:17)
	at org.lwjgl.stb.STBImage.<clinit>(STBImage.java:128)
	at cxo.a(SourceFile:122)
	at cxo.a(SourceFile:98)
	at cxo.a(SourceFile:89)
	at eaz.<init>(SourceFile:29)
	at dbl.a(SourceFile:2409)
	at yt.a(SourceFile:45)
	at eal.a(SourceFile:67)
	at yq.a(SourceFile:33)
	at dbp.a(SourceFile:665)
	at dbl.<init>(SourceFile:439)

-- Initialization --
Details:
Stacktrace:
	at net.minecraft.client.main.Main.main(SourceFile:164)

-- System Details --
Details:
	Minecraft Version: 1.15
	Minecraft Version ID: 1.15
	Operating System: Linux (aarch64) version 6.6.74+rpt-rpi-v8
	Java Version: 1.8.0_442, Amazon.com Inc.
	Java VM Version: OpenJDK 64-Bit Server VM (mixed mode), Amazon.com Inc.
	Memory: 609775752 bytes (581 MB) / 846200832 bytes (807 MB) up to 1908932608 bytes (1820 MB)
	CPUs: 4
	JVM Flags: 2 total; -Xss1M -Xmx2048M
	Launched Version: 1.15
	Backend library: LWJGL version 3.2.1 SNAPSHOT
	Backend API: V3D 7.1.7.0 GL version 3.1 Mesa 24.2.8-1~bpo12+rpt2, Broadcom
	GL Caps: Using framebuffer using OpenGL 3.0
	Using VBOs: Yes
	Is Modded: Probably not. Jar signature remains and client brand is untouched.
	Type: Client (map_client.txt)
	CPU: 5x null
#@!@# Game crashed! Crash report saved to: #@!@# /home/oscar/.config/QuantumLauncher/instances/1.15/.minecraft/crash-reports/crash-2025-04-04_10.59.06-client.txt
```
1.16:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000007f98fb5f18, pid=15122, tid=0x0000007f9f98f180
#
# JRE version: OpenJDK Runtime Environment (8.0_442-b06) (build 1.8.0_442-b06)
# Java VM: OpenJDK 64-Bit Server VM (25.442-b06 mixed mode linux-aarch64 compressed oops)
# Problematic frame:
# C  0x0000007f98fb5f18
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/oscar/.config/QuantumLauncher/instances/1.16/.minecraft/hs_err_pid15122.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/corretto/corretto-8/issues/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

```

**Comment by coderpro1234-2** [comment]
Something of note, is that due to linux updates, the old version now displays properly, which means that your new version might not display properly on an older system. I have not checked the old version of the program with the two different versions of `WGPU_BACKEND`, and had just used it to test what minecraft version work or not.

**Comment by Mrmayman** [comment]
thanks for the information!

That SIGSEGV is weird, can you send the log at `/home/oscar/.config/QuantumLauncher/instances/1.16/.minecraft/hs_err_pid15122.log` ? It might have some useful info

as for that 1.15 problem, i'm not exactly sure how to fix it. But anyway can you check if there's a `liblwjgl_stb.so` file in `/home/oscar/.config/QuantumLauncher/instances/1.16/libraries/natives/`

**Comment by coderpro1234-2** [comment]
Here is the log file:

[hs_err_pid15122.log](https://github.com/user-attachments/files/19658310/hs_err_pid15122.log)

With the library not found error, the file is there, but it downloaded an x86_64 version, not an aarch64 version, so java thinks that it doesn't exist, because it gets an error trying to load it.

**Comment by coderpro1234-2** [comment]
Need anything else?

**Comment by Mrmayman** [comment]
Here is a version with an overhauled library downloader. I believe you tested the v0.4.1 stable release / the version i sent right? This one i am sending is the latest commit with many changes and an attempt to fix this

[ql_arm64_betterjsonstest.zip](https://github.com/user-attachments/files/20590823/ql_arm64_betterjsonstest.zip)

can you try it and see if the game still crashes? note: you may need to create a new instance to receive the improvements

**Comment by coderpro1234-2** [comment]
The new version allows for versions 1.15-1.17 to be downloaded correctly, and launch without issues. I will test the rest of the main version (1.0-1.21.5) to see if they launch, or if they have issues.

**Comment by Mrmayman** [comment]
Alright thanks for the info, do let me know if the other versions launch!

**Comment by coderpro1234-2** [comment]
Minecraft 1.0.0 fails to launch, with the error `java.lang.UnsatisfiedLinkError: /home/oscar/.config/QuantumLauncher/instances/1.0.0/libraries/natives/liblwjgl.so: /home/oscar/.config/QuantumLauncher/instances/1.0.0/libraries/natives/liblwjgl.so: wrong ELF class: ELFCLASS32 (Possible cause: architecture word width mismatch)`. You may need to find a version of the library for aarch64.

**Comment by Mrmayman** [comment]
[quantumlauncherarm64.zip](https://github.com/user-attachments/files/20692765/quantumlauncherarm64.zip)

@coderpro1234-2 can you try this one? I fixed a bug in the library downloader that appeared after overhauling it. 1.0.0 now has the correct libraries

**Comment by coderpro1234-2** [comment]
All of the previous version work (1.0 - 1.21, check in intervals of three), but anything above 1.21 don't run because they can't find files, due to the libraries/native folder being empty.

**Comment by void90user** [comment]
Second one is crash when opening debug log

**Comment by Mrmayman** [comment]
1) there's nothing i can do about that, it's iced library's problem. Try
  - setting UI scale to 1 if you increased it
  - disabling window transparency
  - try other backends. `WGPU_BACKEND=gl` or `WGPU_BACKEND=vulkan` or `ICED_BACKEND=tiny-skia` as env vars put before `cargo run` in the command
  - if all else fails, try the iced 0.14 branch

let me know if any of these workarounds help

2) wdym no zulu java? It's supposed to be there for aarch64, what error are you getting

3) Once again it's an iced issue. Iced just has graphical glitches on raspberry pi at the moment, try the wgpu backend stuff mentioned earlier
